Deprecate and replace `set_sample_data`

unflxw opened the pull request:
This PR is part of [integration-guide#80](https://github.com/appsignal/integration-guide/issues/80).
Fixes #649.

This commit deprecates the `Appsignal.Span.set_sample_data/3` method, replacing it with the `set_params/2`, `set_environment/2`, `set_session_data/2` and `set_custom_data/2` methods.

Before sending the data to the extension, allow-list filtering is applied if the `:phoenix, :filter_parameters` config key is set to `{:keep, keys}`, where `keys` is a list of keys to allow through. Before this change, this filtering would be applied to any data passing through `set_sample_data/3`. Out of the newly implemented functions, only data passing through `set_params/2` has this filtering applied to it.

unflxw commented:
I have no idea why Dialyzer is failing. (@jeffkreeftmeijer thoughts?)

unflxw commented:
TO-DO: Implement support for tags: https://docs.appsignal.com/guides/custom-data/tagging-request.html

tombruijn commented:
> This does not change anything past the Nif module's boundary, which still responds to set_span_sample_data(span, key, value) instead of to some hypothetical set_span_#{key}(span, value). That is what we want from this, yes? Only changing the publicly-exposed interface.

Yes, we only want to provide more user friendly helper methods in the integration. We'll still call `set_span_sample_data` on the Nif.

> I'm not sure if the changelog entry for the :phoenix, :filter_parameters fix should go here or in the appsignal-elixir-phoenix/appsignal-elixir-plug repositories. What fixes it is the fact that the other repositories no longer call set_sample_data (which always does filtering), instead calling set_params (which does the filtering) or set_environment (which does not).

I think it needs to be here because it was fixed here. Even though the fix was moving the behavior to another function. I mean, only updating the other integrations won't fix it.

Actually only upgrading the plug or phoenix package will break things. If users only upgrade the plug or phoenix package it will try to call the `set_params` function that doesn't exist on the older appsignal-elixir package. So I think we need to bump the minimum versions for `appsignal-elixir` in the plug and phoenix packages to be `>= 2.2.8` (assuming the largest bump is a patch bump for the release this is in).

tombruijn commented on lib/appsignal/span.ex:
```diff
@@ -181,28 +181,107 @@ defmodule Appsignal.Span do
+
+  ## Example
+      Appsignal.Tracer.root_span()
+      |> Appsignal.Span.set_headers(%{"X-Request-ID" => "a1b2c3"})
```
```suggestion
      |> Appsignal.Span.set_headers(%{"req_headers.x_request_id" => "a1b2c3"})
```
I think this matches the format of request headers in plug?

My test apps report this format: `req_headers.accept`

tombruijn reviewed (approved):
Very nice 👏

unflxw commented:
Update: fixed the dialyzer warning by adding `:ex_unit` to the apps that it analyses. This makes me wonder if I should not be calling `assert` from the `Appsignal.Test.Span`? I saw that the `Appsignal.Test.Wrapper` was already importing `ExUnit.Assertions`, so I thought it would be okay, but now I'm not so sure.

jeffkreeftmeijer commented:
Using `assert` in the test modules adds a dependency to ExUnit at runtime, which Dialyzer complains about. That’s because the Test modules are exported to be used in testing the Plug and Phoenix packages. Could you move the assertions to the test suite instead? Other than that, this patch is great. :)

jeffkreeftmeijer reviewed (changes requested):
https://github.com/appsignal/appsignal-elixir/pull/713#issuecomment-966097152

unflxw commented:
@jeffkreeftmeijer Thanks! I had moved those assertions to `Test.Span` precisely so I could reuse them on the Plug and Phoenix integrations. I have now duplicated them across the repos and removed the Dialyzer exception. It would be nice if there was a way to share test components (including assertions!) across the three repos, though.

Regarding the versioning kerfuffle (ping @tombruijn), I have done the following:
- Increased the version number of `:appsignal` to 2.2.8, and modified the changes in `:appsignal_plug` so that it targets that version in its dependency to `:appsignal`.
- Increased the version number of `:appsignal_plug` to 2.0.9, and modified the changes in `:appsignal_phoenix` so that t targes that version in its dependency to `:appsignal_plug`.

I have aimed for the minimum set of changes to get the dependencies right after merge, and as such, I have not increased the version number of `:appsignal_phoenix` -- in practice, we probably will want to publish a new version of `:appsignal_phoenix` with an increased version number as well, so the affected users can target it.

tombruijn commented on mix.exs:
```diff
@@ -12,7 +12,7 @@ defmodule Appsignal.Mixfile do
 
   @source_url "https://github.com/appsignal/appsignal-elixir"
-  @version "2.2.7"
+  @version "2.2.8"
```
I don't quite understand from your comment about the version kerfuffle, but this version bump is done by mono, you don't need to update it in this PR, or at least not in the merged version.

unflxw replied:
Right! In that case, is it okay for the pull requests on the Phoenix and Plug repositories to point to a not-yet-released version of `appsignal-elixir`?

tombruijn replied:
Yes :)

unflxw commented:
TODO: don't take a span as an argument in the new helpers; instead, always operate on Appsignal.Tracer.root_span -- [see Basecamp conversation](https://3.basecamp.com/4738529/buckets/18166704/messages/4360963269#__recording_4361563202)

tombruijn commented on lib/appsignal/utils/map_filter.ex:
```diff
@@ -1,4 +1,7 @@
 defmodule Appsignal.Utils.MapFilter do
+  # TO-DO: Move this module to the Phoenix package, affter
```
```suggestion
  # TODO: Move this module to the Phoenix package, after
```
It's more common to write it like this and easier to find back. I only look for "TODO" or "todo".

tombruijn reviewed (approved):
See also my comments on the phoenix PR https://github.com/appsignal/appsignal-elixir-phoenix/pull/36#discussion_r756754296

tombruijn commented:
I found a problem with this as described in: https://github.com/appsignal/integration-guide/issues/80#issuecomment-979318138

unflxw commented:
We're going back to the drawing board with this whole sample data helpers thing. This implementation makes assumptions that don't always hold true. For the time being, we'll implement a separate fix for #649 and leave the rest of this change unmerged.
